src/main/java/io/nats/client/impl/Headers.java

```diff
@@ -36,21 +36,47 @@ public class Headers {
 
 	private final Map<String, List<String>> valuesMap;
 	private final Map<String, Integer> lengthMap;
+	private final boolean readOnly;
 	private byte[] serialized;
 	private int dataLength;
 
 	public Headers() {
-		valuesMap = new HashMap<>();
-		lengthMap = new HashMap<>();
+		this(null, false, null);
 	}
 
 	public Headers(Headers headers) {
-		this();
+		this(headers, false, null);
+	}
+
+	public Headers(Headers headers, boolean readOnly) {
+		this(headers, readOnly, null);
+	}
+
+	public Headers(Headers headers, boolean readOnly, String[] keysNotToCopy) {
+		Map<String, List<String>> tempValuesMap = new HashMap<>();
+		Map<String, Integer> tempLengthMap = new HashMap<>();
 		if (headers != null) {
-			valuesMap.putAll(headers.valuesMap);
-			lengthMap.putAll(headers.lengthMap);
+			tempValuesMap.putAll(headers.valuesMap);
+			tempLengthMap.putAll(headers.lengthMap);
 			dataLength = headers.dataLength;
-			serialized = null;
+			if (keysNotToCopy != null) {
+				for (String key : keysNotToCopy) {
+					if (key != null) {
+						if (tempValuesMap.remove(key) != null) {
+							dataLength -= tempLengthMap.remove(key);
+						}
+					}
+				}
+			}
+		}
+		this.readOnly = readOnly;
+		if (readOnly) {
+			valuesMap = Collections.unmodifiableMap(tempValuesMap);
```

Comment:
unmodifiableMap throws an `UnsupportedOperationException` when `computeIfAbsent` is called which is happening in the `_add` method.

Reply:
@sdklib Are you trying to modify the headers after the message is created? This PR was intended to prevent this.

Reply:
Yup, we're using AOP to add a correlation id for proper logging across microservices to the headers.

After updating to the newest version, some unit tests fail, due to readOnly being true and the AOP Pointcut failing to add the correlation ID to the header.

The fix was easy, I was just wondering if the response should really be `UnsupportedOperationException` or maybe better returning a `Headers.class` clone that doesn't have the add method in the first place. :-)

Reply:
Feared so, never the less, thank you so much! 👍